### Introducción al Proyecto de Análisis de Tendencias de Videos Publicitarios

En la agencia de publicidad **Sterling & Draper**, trabajo como analista de videos publicitarios y me especializo en analizar las tendencias de videos en YouTube para ayudar a tomar decisiones estratégicas de mercadotecnia. El objetivo principal de mi trabajo es identificar patrones en el contenido que se vuelve viral, ya que esto puede indicar las preferencias de los consumidores y ayudar a definir campañas publicitarias más efectivas.

El análisis de tendencias se enfoca en los **videos de YouTube** que se encuentran en la sección de tendencias, donde los usuarios descubren contenido popular. Cada video está vinculado a una **categoría específica** (como entretenimiento, música, noticias, política, etc.), y a una **región geográfica** (como Estados Unidos, Francia, Japón, etc.), además de la **fecha** en que el video comienza a ser tendencia.

Sin embargo, el proceso de revisión manual de las tendencias es extenso y requiere mucho tiempo. Esto se vuelve especialmente complicado cuando los nuevos empleados, como Melanie y Ashok, preguntan de manera recurrente sobre las categorías que estuvieron en tendencia la semana pasada, cómo se distribuyeron estas tendencias entre diversas regiones y qué categorías fueron particularmente populares.

Para agilizar este proceso, decido automatizar el análisis creando un **dashboard interactivo** que permita a los gerentes de planificación de videos publicitarios acceder rápidamente a los datos históricos de tendencias de videos, organizados por categorías y regiones. Este dashboard facilitará la toma de decisiones y proporcionará información en tiempo real sobre el comportamiento de las tendencias de videos en YouTube.

#### Requisitos Técnicos del Proyecto

- **Objetivo de negocio:** Analizar el historial de tendencias de videos en YouTube para optimizar las estrategias publicitarias.
  
- **Frecuencia de uso del dashboard:** El dashboard se usará al menos una vez al día, ya que se actualizará automáticamente con datos nuevos cada 24 horas.

- **Usuarios objetivo:** El dashboard está destinado a los **gerentes de planificación de videos publicitarios**, quienes tomarán decisiones sobre campañas basadas en los datos presentados.

- **Contenido del dashboard:** El dashboard deberá contener tres tipos de datos clave:
  - **Tendencias pasadas de videos** organizadas por fechas y categoría.
  - **Distribución de tendencias de videos** por región geográfica.
  - **Correspondencia entre categorías y países** para entender las categorías más populares en cada región.

- **Parámetros para agrupar los datos:**
  - **Fecha de la tendencia.**
  - **Categoría del video.**
  - **País o región.**

- **Datos del dashboard:**
  1. **Historial de tendencias:** Información organizada por semana, incluyendo tanto números absolutos de videos como su proporción porcentual de aparición en tendencias.
  2. **Eventos por país:** Valores relativos de eventos, como el porcentaje de videos por categoría en cada región.
  3. **Correspondencia entre categorías y países:** Una tabla que muestra cuántos videos de cada categoría están en tendencia en cada región, en valores absolutos.

- **Fuente de datos:** Los ingenieros de la base de datos proporcionarán una tabla de agregación llamada `trending_by_time` que contiene los siguientes campos:
  - **record_id:** La clave primaria del registro.
  - **region:** País o región geográfica donde se encuentra el video.
  - **trending_date:** La fecha y hora en que el video comienza a ser tendencia.
  - **category_title:** La categoría a la que pertenece el video (como entretenimiento, música, etc.).
  - **videos_count:** El número de videos que estuvieron en la sección de tendencias en un día determinado.
  
- **Actualización de los datos:** Los datos serán actualizados una vez cada 24 horas, a la media noche UTC.

#### Visualización y Funcionalidades del Dashboard

El dashboard se organizará en varias secciones, con gráficos e interacciones fáciles de usar, para que Melanie, Ashok y otros gerentes de publicidad puedan obtener rápidamente respuestas a sus preguntas más frecuentes. Los gráficos incluirán:

1. **Gráficos de tendencias históricas**: 
   - Un gráfico de barras que muestra los videos que estuvieron en tendencias por categoría, ordenado por fecha.
   - Un gráfico de líneas que muestra la proporción porcentual de cada categoría en comparación con el total de videos de tendencias.

2. **Distribución de videos por región**:
   - Un gráfico de mapas o barras apiladas que muestra cómo se distribuyen las categorías de videos en las diferentes regiones geográficas.

3. **Tabla de correspondencia entre categorías y países**:
   - Una tabla detallada que indica cuántos videos de cada categoría estuvieron en tendencia en cada región, con filtros para facilitar la búsqueda de información.

#### Beneficios del Proyecto

Este proyecto no solo agiliza el proceso de análisis de tendencias de videos, sino que también ayuda a **Sterling & Draper** a tomar decisiones informadas sobre campañas publicitarias al proporcionar una visión clara y accesible de las tendencias de YouTube. El dashboard permitirá a los gerentes detectar patrones en las preferencias de los usuarios, ajustar sus estrategias de contenido y dirigir sus esfuerzos de marketing hacia los videos y categorías más populares en las regiones clave.

In [1]:
####################
## INICIALIZACION ##
####################

#Importar librerias 
import pandas as pd
import numpy as np
from datetime import datetime, timedelta # Filtrar los datos de la semana pasada
import plotly.graph_objects as go
import plotly.express as px


In [2]:
##################
## Cargar datos ##
##################
df = pd.read_csv('/Users/fransuryhenaosepulveda/Documents/Sprint 12/trending_by_time.csv')

# Verificar la información
df.info()
display(df.head())
df.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12343 entries, 0 to 12342
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   record_id       12343 non-null  int64 
 1   region          12343 non-null  object
 2   trending_date   12343 non-null  object
 3   category_title  12343 non-null  object
 4   videos_count    12343 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 482.3+ KB


record_id  region               trending_date    category_title  \
0          1  France  2017-11-14 00:00:00.000000  Autos & Vehicles   
1          2  France  2017-11-15 00:00:00.000000  Autos & Vehicles   
2          3  France  2017-11-16 00:00:00.000000  Autos & Vehicles   
3          4  France  2017-11-17 00:00:00.000000  Autos & Vehicles   
4          5  France  2017-11-18 00:00:00.000000  Autos & Vehicles   

   videos_count  
0             8  
1             2  
2             6  
3             8  
4             4

np.int64(0)

Los datos se han cargado correctamente y puedes ver la estructura del DataFrame `df` con la información básica sobre las columnas y su tipo de datos. No temenos datos duplicados.

### Descripción de las columnas:
1. **`record_id`**: Identificador único del registro (tipo de dato: `int64`).
2. **`region`**: País o región donde el video estuvo en tendencia (tipo de dato: `object`, es decir, cadena de texto).
3. **`trending_date`**: Fecha y hora cuando el video se hizo tendencia (tipo de dato: `object`, vamos a convertir a `datetime`).
4. **`category_title`**: Categoría del video (tipo de dato: `object`, cadena de texto).
5. **`videos_count`**: Número de videos en tendencia para esa fecha y región (tipo de dato: `int64`).

### Próximos pasos sugeridos:
1. **Convertir `trending_date` a formato datetime**:
   Como `trending_date` está en formato de cadena (`object`), sería útil convertirlo a un tipo `datetime` para facilitar el análisis temporal, como seleccionar por fechas o agrupar por semana o mes.

2. **Explorar los datos**:
   Puedes empezar a explorar las tendencias de videos por categoría, región y fecha. Por ejemplo, podrías analizar la cantidad de videos en tendencia por categoría a lo largo del tiempo o en distintas regiones.

3. **Análisis básico**:
   Algunas operaciones iniciales que podrías realizar:
   - Ver las categorías y regiones únicas con `df['category_title'].unique()` y `df['region'].unique()`.
   - Calcular estadísticas descriptivas sobre los conteos de videos con `df['videos_count'].describe()`.

In [3]:
############################
## PROCESAMIENTO DE DATOS ##
############################

# Convertir `trending_date` a formato datetime
df['trending_date'] = pd.to_datetime(df['trending_date'])

# Crear una columna con el número del mes
df['month'] = df['trending_date'].dt.month

# Crear una nueva columna con el año
df['year'] = df['trending_date'].dt.year

# Verificar la estructura del DataFrame
df.info()

# Mostrar los primeros registros
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12343 entries, 0 to 12342
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   record_id       12343 non-null  int64         
 1   region          12343 non-null  object        
 2   trending_date   12343 non-null  datetime64[ns]
 3   category_title  12343 non-null  object        
 4   videos_count    12343 non-null  int64         
 5   month           12343 non-null  int32         
 6   year            12343 non-null  int32         
dtypes: datetime64[ns](1), int32(2), int64(2), object(2)
memory usage: 578.7+ KB


record_id  region trending_date    category_title  videos_count  month  \
0          1  France    2017-11-14  Autos & Vehicles             8     11   
1          2  France    2017-11-15  Autos & Vehicles             2     11   
2          3  France    2017-11-16  Autos & Vehicles             6     11   
3          4  France    2017-11-17  Autos & Vehicles             8     11   
4          5  France    2017-11-18  Autos & Vehicles             4     11   

   year  
0  2017  
1  2017  
2  2017  
3  2017  
4  2017

In [4]:
#Ver las categorías únicas 
df['category_title'].unique()

array(['Autos & Vehicles', 'Comedy', 'Education', 'Entertainment',
       'Film & Animation', 'Gaming', 'Howto & Style', 'Movies', 'Music',
       'News & Politics', 'People & Blogs', 'Pets & Animals',
       'Science & Technology', 'Shows', 'Sports', 'Trailers',
       'Travel & Events', 'Nonprofits & Activism'], dtype=object)

In [5]:
#regiones únicas
df['region'].unique()

array(['France', 'India', 'Japan', 'Russia', 'United States'],
      dtype=object)

In [6]:
#Calcular estadísticas descriptivas sobre los conteos de videos
df['videos_count'].describe()

count    12343.000000
mean        27.545167
std         29.793491
min          2.000000
25%          8.000000
50%         18.000000
75%         36.000000
max        220.000000
Name: videos_count, dtype: float64

## Análisis sobre los datos

Los datos abarcan las siguientes cinco regiones: France, India, Japan, Russia, y United States. Esto sugiere un análisis global con regiones de diferentes continentes.

- Estadísticas descriptivas de los videos:
* Cantidad total de registros: 12,343 (número total de observaciones).
* Promedio de videos por registro: 27.55 videos, lo que indica que, en promedio, cada registro tiene alrededor de 28 videos.
* Desviación estándar: 29.79, lo que implica una variabilidad significativa en la cantidad de videos entre los registros.
* Mínimo y máximo: El registro más bajo tiene 2 videos, mientras que el más alto tiene 220 videos.

- Percentiles:
* 25%: Un cuarto de los registros tienen 8 videos o menos.
* 50% (mediana): La mitad de los registros tienen 18 videos o menos.
* 75%: Tres cuartos de los registros tienen 36 videos o menos.

- Variabilidad:
El rango amplio (2 a 220) y la desviación estándar alta muestran que hay regiones o categorías con un número de videos muy superior al promedio.

# Conclusión general
La mayoría de los registros tienen una cantidad moderada de videos (entre 8 y 36), pero existen valores atípicos con cantidades excepcionalmente altas (hasta 220 videos), lo cual puede estar relacionado con tendencias virales o categorías populares en ciertas regiones.

In [7]:
######################################
## CONFIGURACION DE VISUALIZACIONES ##
######################################

#Definir paletas de colores
color_fondo = "#f2e8cf"
color_fill = ["#16425b","#3a7ca5","#81c3d7","#ccc9dc","#F79489","#F8AFA6","#FADCD9","#F9F1F0"]
color_texto = ["#424B54","white"]
color_borde = "#424B54"

# Definir formatos a usar en todos los graficos
form_etiquetas = dict(
    font = dict(size = 15)
)

form_ticks = dict(
    tickfont = dict(size = 15)
)

form_leyendas = dict(
    orientation = "h", 
    y = 1, 
    yanchor = "bottom",
    font = dict(size = 15)
)

In [8]:
# Calcular proporciones porcentuales
df['total_videos_per_date'] = df.groupby('trending_date')['videos_count'].transform('sum')
df['percentage'] = (df['videos_count'] / df['total_videos_per_date']) * 100

# Gráfico de Barras
fig_barras = go.Figure()

# Añadir trazas por categoría
for category in df['category_title'].unique():
    df_cat = df[df['category_title'] == category]
    fig_barras.add_trace(
        go.Bar(
            x=df_cat['trending_date'],
            y=df_cat['videos_count'],
            name=category
        )
    )

fig_barras.update_layout(
    title='Videos en Tendencias por Categoría (Ordenado por Fecha)',
    xaxis_title='Fecha',
    yaxis_title='Cantidad de Videos',
    barmode='stack',
    template='plotly',
    plot_bgcolor='#f2e8cf'
)

# Gráfico de Líneas
fig_lineas = go.Figure()

# Añadir trazas por categoría
for category in df['category_title'].unique():
    df_cat = df[df['category_title'] == category]
    fig_lineas.add_trace(
        go.Scatter(
            x=df_cat['trending_date'],
            y=df_cat['percentage'],
            mode='lines+markers',
            name=category
        )
    )

fig_lineas.update_layout(
    title='Proporción Porcentual de Cada Categoría en Tendencias',
    xaxis_title='Fecha',
    yaxis_title='Porcentaje (%)',
    template='plotly',
    plot_bgcolor='#f2e8cf'
)

# Mostrar gráficos
fig_barras.show()
fig_lineas.show()

## Gráfico de Barras Apiladas (Cantidad de Videos en Tendencias por Categoría)
Observaciones:

La cantidad total de videos en tendencias muestra cierta estacionalidad o variaciones a lo largo del tiempo.
Hay picos y caídas regulares, que podrían estar relacionados con eventos globales, lanzamientos de contenido, o cambios en el algoritmo de tendencias.

Algunas categorías como "Entertainment" (la de color predominante púrpura) parecen dominar el número total de videos en tendencias, otras categorías como "Music" y "Comedy" tienen contribuciones significativas, pero no superan a "Entertainment".

Categorías como "Nonprofits & Activism", "Travel & Events" y "Trailers" tienen una representación mínima en las tendencias.

# Conclusión:

El gráfico muestra un dominio claro de ciertas categorías como "Entertainment" y "Music", mientras que las categorías más pequeñas tienen poca representación. Esto podría indicar un sesgo en los videos en tendencias hacia contenido más popular o comercial.

## Gráfico de Líneas (Proporción Porcentual de Cada Categoría en Tendencias)

Algunas categorías, como "Entertainment" y "Music", tienen una proporción porcentual relativamente estable, aunque con ligeras fluctuaciones.

Algunas categorías muestran una tendencia a la baja en su proporción porcentual con el tiempo (por ejemplo, "Film & Animation").

Hay una alta diversidad de categorías, pero las categorías principales dominan la mayor parte de las tendencias. Para las categorías menores, los porcentajes parecen más ruidosos, posiblemente debido a la baja cantidad de videos en esas categorías.

# Conclusión:

El gráfico indica que, aunque las tendencias permiten cierta diversidad de categorías, el peso relativo de las categorías principales (como "Entertainment" y "Music") sigue siendo dominante. Esto podría reflejar las preferencias del público o las dinámicas del algoritmo de tendencias.

## Conclusión General sobre Ambos Gráficos

Las categorías de contenido popular como "Entertainment" y "Music" dominan consistentemente en cantidad y proporción.

Aunque hay representación de categorías más pequeñas, su impacto es mínimo, lo que podría indicar un sesgo hacia contenido comercial o ampliamente consumido.

La estacionalidad o eventos globales parecen influir en los picos y caídas en la cantidad de videos en tendencias.

In [9]:
# Gráfico de barras para la cantidad de videos por categoría y región

category_counts = df.groupby(['region', 'category_title'])['videos_count'].sum().reset_index()

# Crear el gráfico de barras
fig = px.bar(
    category_counts,
    x="region",  # Eje X: Región
    y="videos_count",  # Eje Y: Número de videos
    color="category_title",  # Agrupar por categoría
    barmode="stack",  # Barras apiladas
    title="Número de Videos en Tendencia por Región y Categoría",
    labels={"videos_count": "Número de Videos", "region": "Región", "category_title": "Categoría"},
    color_discrete_sequence=px.colors.qualitative.Set2  # Paleta de colores
)

fig.update_layout(
    plot_bgcolor = color_fondo,
    xaxis = form_ticks,
    yaxis = dict(),
    title='Cantidad total de videos en tendencia por categoría y región',
    
    xaxis_title='Categoría',
    yaxis_title='Número de videos',
    autosize = False,
    width = 1000,
    height = 450
)

fig.show()

# Cantidad total de videos en tendencia por categoría y región, se pueden destacar las siguientes conclusiones:

1. Dominancia de Entretenimiento: La categoría Entretenimiento tiene un peso notable en todas las regiones, representando una porción significativa del total de videos en tendencia.
2. Variaciones Regionales: En Francia, India, y Estados Unidos, la categoría de Música también es predominante, mientras que en Japón y Rusia, parece haber mayor equilibrio entre categorías como Educación, Comedia, y Noticias y Política.
3. Categorías Menos Populares: Categorías como Autos y Vehículos, Películas, y Estilo tienen una representación mucho menor en todas las regiones, lo que sugiere que no generan tanto contenido en tendencia.
4. Diferencias Culturales y de Preferencias: En Rusia, la categoría Noticias y Política parece ser más popular que en regiones como India o Estados Unidos, reflejando tal vez un interés más alto por contenidos de actualidad. En Japón, categorías como Educación y Ciencia y Tecnología tienen una presencia significativa, lo que podría estar vinculado a intereses culturales específicos.

# Conclusión General
Este gráfico resalta cómo las tendencias de videos están influenciadas por preferencias regionales y culturales. Aunque categorías universales como Entretenimiento y Música dominan en la mayoría de las regiones, existen diferencias claras que pueden ser útiles para ajustar estrategias de contenido según la audiencia objetivo.

In [10]:
# Crear un rango de fechas (semanas disponibles)
fechas = pd.date_range(start=df['trending_date'].min(), end=df['trending_date'].max(), freq='W-MON')

# Almacenar resultados por semana
resultados = []

for fecha in fechas:
    inicio_semana = fecha - timedelta(days=7)
    fin_semana = fecha - timedelta(days=1)
    
    # Filtrar datos de la semana
    df_semana = df[(df['trending_date'] >= inicio_semana) & (df['trending_date'] <= fin_semana)]
    
    # Agrupar por categoría y región para toda la semana
    categorias_regiones = df_semana.groupby(['category_title', 'region'])['videos_count'].sum().reset_index()
    
    # Análisis para cada región
    regiones = df['region'].unique()  # Obtener todas las regiones únicas
    categorias_por_region = {}

    for region in regiones:
        categorias_region = df_semana[df_semana['region'] == region].groupby('category_title')['videos_count'].sum().reset_index()
        categorias_por_region[region] = categorias_region  # Guardar resultados para la región
    
    # Guardar los resultados de la semana
    resultados.append({
        'semana_inicio': inicio_semana,
        'semana_fin': fin_semana,
        'categorias_regiones': categorias_regiones,
        'categorias_por_region': categorias_por_region
    })

# Acceso a los datos procesados por semana y región
for resultado in resultados:
    print(f"Semana {resultado['semana_inicio']} a {resultado['semana_fin']}")
    print("Categorías y Regiones (Generales):")
    print(resultado['categorias_regiones'].head())
    
    print("\nAnálisis por Región:")
    for region, categorias_region in resultado['categorias_por_region'].items():
        print(f"- Región: {region}")
        print(categorias_region.head())

Semana 2017-11-13 00:00:00 a 2017-11-19 00:00:00
Categorías y Regiones (Generales):
     category_title         region  videos_count
0  Autos & Vehicles         France            38
1  Autos & Vehicles          India            12
2  Autos & Vehicles         Russia           124
3  Autos & Vehicles  United States            40
4            Comedy         France           328

Análisis por Región:
- Región: France
     category_title  videos_count
0  Autos & Vehicles            38
1            Comedy           328
2         Education            42
3     Entertainment           548
4  Film & Animation           122
- Región: India
     category_title  videos_count
0  Autos & Vehicles            12
1            Comedy           212
2         Education            92
3     Entertainment           978
4  Film & Animation            78
- Región: Japan
Empty DataFrame
Columns: [category_title, videos_count]
Index: []
- Región: Russia
     category_title  videos_count
0  Autos & Vehicles       

In [11]:
# Crear el gráfico
fig = go.Figure()

# Iterar sobre las categorías únicas
categorias = df_semana['category_title'].unique()
for categoria in categorias:
    data_categoria = df_semana[df_semana['category_title'] == categoria]
    fig.add_trace(
        go.Bar(
            x=data_categoria['region'],
            y=data_categoria['videos_count'],
            name=categoria,
            text=data_categoria['videos_count'],
            textposition='auto',
            hoverinfo="x+y+name"
        )
    )

# Configurar el diseño del gráfico
fig.update_layout(
    title="Videos en Tendencias por Categoría y Región (Semana del 13 al 19 de Noviembre de 2017)",
    xaxis_title="Regiones",
    yaxis_title="Cantidad de Videos",
    barmode="stack",
    template="plotly_white",
    legend_title="Categorías"
)

# Mostrar el gráfico
fig.show()

## Conclusión sobre los resultados de la semana 2017-11-13 a 2017-11-19:

Las categorías más representadas en las regiones analizadas son Autos & Vehicles, Comedy, Education, Entertainment, y Film & Animation.
La categoría Entertainment muestra consistentemente altos conteos en varias regiones, lo que indica su popularidad global durante esta semana.

# Por región:
* Francia: 
- Entertainment (548 videos) lidera en esta región, mostrando una clara preferencia por contenidos de entretenimiento.
- Comedy (328 videos) ocupa un segundo lugar, indicando un interés significativo en contenido humorístico.

* India: Presenta el mayor interés en Entertainment (978 videos), superando significativamente a otras categorías.
También destaca un interés relevante en Education (92 videos), lo que podría reflejar el consumo de contenido educativo junto con entretenimiento.

* Rusia:
- Autos & Vehicles (124 videos) tiene una representación notable en Rusia, superando considerablemente otras regiones en esta categoría. Sin embargo, Entertainment (710 videos) continúa siendo la categoría dominante.

* Japón: Muestra interés en categorías variadas, como Education (64 videos), Gaming (112 videos), y Film & Animation (214 videos). Esto sugiere un público diversificado en sus intereses.

Patrones Notables:
La categoría Entertainment domina globalmente, destacando como una tendencia universal,  las categorías secundarias, como Comedy, Film & Animation, y Autos & Vehicles, muestran variaciones regionales que podrían estar influenciadas por preferencias culturales.

Autos & Vehicles tiene una presencia significativa en Rusia, lo que podría relacionarse con el interés en la industria automotriz local.

Los datos podrían orientar la producción de contenido específico por región. Por ejemplo, contenido humorístico y de entretenimiento para Francia, y más contenido educativo para India.
También sugieren oportunidades para industrias específicas, como automotriz en Rusia y videojuegos en Japón.

In [12]:
fig_lines = go.Figure()

# Iterar por categoría para añadir líneas
for category in df['category_title'].unique():
    porcentaje_semanal = []

    # Calcular proporción semanal por categoría
    for resultado in resultados:
        semana_total = resultado['categorias_regiones']['videos_count'].sum()
        categoria_total = resultado['categorias_regiones'][resultado['categorias_regiones']['category_title'] == category]['videos_count'].sum()
        porcentaje_semanal.append((categoria_total / semana_total) * 100 if semana_total > 0 else 0)
    
    # Fechas de las semanas
    semanas = [f"{res['semana_inicio'].strftime('%Y-%m-%d')} - {res['semana_fin'].strftime('%Y-%m-%d')}" for res in resultados]
    
    # Añadir traza
    fig_lines.add_trace(
        go.Scatter(
            x=semanas,
            y=porcentaje_semanal,
            mode='lines+markers',
            name=category,
            hoverinfo="x+y+name"
        )
    )

# Configurar el diseño del gráfico
fig_lines.update_layout(
    title="Proporción Porcentual de Cada Categoría en Tendencias (Semanal)",
    xaxis_title="Semanas",
    yaxis_title="Porcentaje (%)",
    template="plotly_white"
)

fig_lines.show()

## Análisis de la Proporción Porcentual de Cada Categoría en Tendencias (Semanal)

La categoría "Entertainment" tiene la proporción más alta durante todo el período analizado, alcanzando alrededor del 30% del total de videos en tendencias semanalmente. Esto indica que los videos de entretenimiento son consistentemente los más populares y predominantes en las tendencias.

Las categorías como "Comedy", "Film & Animation" y "Gaming" muestran proporciones más bajas, pero son relativamente estables a lo largo del tiempo, sin cambios abruptos en su representación.

Categorías como "Education", "Autos & Vehicles" y "Howto & Style" tienen una representación porcentual mucho menor (generalmente por debajo del 5%), indicando un interés más limitado o una menor prioridad en las tendencias.

No se observan cambios significativos en las proporciones a lo largo del tiempo, lo que sugiere que los patrones de consumo en las tendencias son consistentes y podrían estar influenciados por las preferencias del público o por los algoritmos de la plataforma.

## Conclusión General
El gráfico demuestra una clara dominancia de contenido de "Entertainment", seguido por categorías como "Comedy" y "Film & Animation". Aunque hay diversidad en las categorías representadas, la mayoría de ellas tienen una participación relativamente pequeña. Esto podría reflejar un sesgo hacia contenido ampliamente atractivo o comercial, dejando a las categorías más específicas con menos visibilidad en las tendencias.

In [13]:
# Crear un DataFrame consolidado con los datos de videos por región
consolidado_regiones = pd.concat([resultado['categorias_regiones'] for resultado in resultados]).groupby('region')['videos_count'].sum().reset_index()

# Crear el gráfico de pastel
fig = go.Figure(
    data=[go.Pie(
        labels=consolidado_regiones['region'],  # Etiquetas (regiones)
        values=consolidado_regiones['videos_count'],  # Valores (número de videos)
        hole=0.4,  # Para convertirlo en gráfico de dona
        textinfo='percent+label',  # Mostrar porcentaje y etiqueta
        marker=dict(colors=color_fill)  # Usar la paleta de colores definida
    )]
)

# Personalizar el diseño con configuración global
fig.update_layout(
    plot_bgcolor=color_fondo,  # Fondo definido
    title=dict(
        text="Tendencias de Videos por País (Relativo)",
        font=dict(size=20)  # Tamaño de título
    ),
    legend=dict(
        title="Regiones",
        orientation=form_leyendas['orientation'],  # Orientación horizontal
        y=form_leyendas['y'],  # Posición vertical
        yanchor=form_leyendas['yanchor'],  # Anclaje vertical
        font=form_leyendas['font']  # Estilo de fuente
    ),
    autosize=False,
    width=700,  # Ancho fijo
    height=500  # Altura fija
)

# Mostrar el gráfico
fig.show()

## Conclusión sobre el gráfico de tendencias de videos por país (relativo):

El gráfico muestra una distribución relativamente equilibrada entre las regiones, lo que indica que todas tienen una representación significativa en términos de videos tendencia.
Sin embargo, Estados Unidos (23.8%) tiene la mayor proporción, lo que refleja su influencia considerable en la producción o consumo de contenido en tendencia.

- Estados Unidos (23.8%): Lidera entre las regiones analizadas, posiblemente debido a su gran volumen de contenido y diversidad de usuarios.
- Francia (22.2%) y Rusia (21.7%): Siguen muy de cerca, lo que destaca la fuerte presencia de estas regiones en las tendencias globales. Rusia, en particular, podría estar impulsada por una alta producción en categorías específicas como autos o entretenimiento.
- India (21.6%): Su proporción también es significativa, lo que puede estar relacionado con su vasto mercado digital y el creciente consumo de contenido en plataformas como YouTube.
- Japón (10.7%): Tiene la menor proporción entre las regiones, aunque su participación sigue siendo importante. Esto puede reflejar preferencias más específicas en contenido o una menor producción relativa en comparación con otras regiones.

La representación global de contenido tendencia es amplia, pero hay un dominio claro de regiones específicas como Estados Unidos y Europa.
Japón, aunque tiene un menor porcentaje, podría enfocarse en nichos de contenido que no necesariamente tienen un impacto global, pero son altamente relevantes localmente.

Los creadores de contenido y plataformas pueden identificar oportunidades para personalizar el contenido basado en estas proporciones.
Las regiones con menor representación relativa, como Japón, pueden representar mercados específicos o áreas donde hay espacio para un mayor crecimiento en tendencias globales.

In [14]:
# Crear una figura
fig = go.Figure()

# Añadir trazas para cada región
regiones = categorias_regiones['region'].unique()
for region in regiones:
    df_region = categorias_regiones[categorias_regiones['region'] == region]
    fig.add_trace(
        go.Bar(
            x=df_region['category_title'],
            y=df_region['videos_count'],
            name=region
        )
    )

# Personalizar el diseño
fig.update_layout(
    plot_bgcolor=color_fondo,  # Fondo del gráfico
    xaxis=dict(  # Configuración del eje X
        title="Categoría",
        tickangle=-45,
        tickfont=form_ticks['tickfont']
    ),
    yaxis=dict(  # Configuración del eje Y
        title="Cantidad de Videos",
        tickfont=form_ticks['tickfont']
    ),
    title=dict(  # Configuración del título
        text="Distribución de Videos por Categoría y Región (Semana Anterior)",
        font=form_etiquetas['font']
    ),
    legend=dict(  # Configuración de la leyenda
        title="Región",
        orientation=form_leyendas['orientation'],
        y=form_leyendas['y'],
        yanchor=form_leyendas['yanchor'],
        font=form_leyendas['font']
    ),
    barmode="stack",  # Barras apiladas
    autosize=False,  # Dimensiones ajustadas manualmente
    width=900,
    height=600
)

# Mostrar la figura
fig.show()

## Análisis de la Distribución de Videos por Categoría y Región (Semana Anterior)

La categoría "Entertainment" es claramente dominante en todas las regiones, con un número significativamente mayor de videos en tendencia, especialmente en India y Estados Unidos, esto confirma que el contenido de entretenimiento tiene un atractivo global y lidera en tendencias.

Algunas categorías tienen un comportamiento notable según la región:
- India tiene una gran cantidad de videos en las categorías de "Entertainment" y "Music", mostrando un fuerte enfoque en contenido culturalmente popular.
- Estados Unidos también lidera en "Entertainment", pero también tiene un alto número en categorías como "News & Politics" y "Comedy".
- Francia y Rusia tienen representaciones más equilibradas entre "Entertainment", "Music", y otras categorías.
- Japón muestra cierta preferencia en categorías como "Film & Animation", lo que podría estar relacionado con la popularidad de contenido animado en esa región.

Categorías como "Autos & Vehicles", "Nonprofits & Activism", y "Science & Technology" tienen poca representación en todas las regiones, lo que indica un menor interés global en este tipo de contenido.

"Music" y "People & Blogs", estas categorías tienen una buena representación en varias regiones, especialmente en India y Estados Unidos, mostrando que el contenido relacionado con música y blogs personales tiene un atractivo notable.

# Conclusión General
El gráfico revela que el contenido de "Entertainment" domina las tendencias globalmente, mientras que las preferencias secundarias varían según las regiones. "Music" es fuerte en países culturalmente ricos en música como India, mientras que "Film & Animation" tiene mayor relevancia en Japón. Por otro lado, categorías más técnicas o específicas tienen menor representación, lo que podría reflejar el enfoque principal de las plataformas en el entretenimiento y contenido generalista.

In [15]:
# Agrupar los datos por categoría y región y sumar la cantidad de videos
table = df.groupby(['category_title', 'region'])['videos_count'].sum().unstack(fill_value=0).reset_index()

# Extraer encabezados y datos
header_values = ["<b>Categoría</b>"] + list(table.columns[1:])  # Excluir la columna duplicada de 'category_title'
cell_values = [table['category_title']] + [table[col].values for col in table.columns[1:]]

# Generar visualización
fig = go.Figure(
    data=go.Table(
        header=dict(
            values=header_values,
            fill_color=color_fill[:2],
            font_color=color_texto[1],
            font_size=20,
            align="center"
        ),
        cells=dict(
            values=cell_values,
            format=["", "", ","],
            fill_color=color_fondo,
            height=25,
            font_size=18,
            font_color=color_texto[0],
            align="center"
        ),
        columnwidth=[50] + [22] * (len(table.columns) - 1)
    )
)

fig.update_layout(
    autosize=False,
    width=1000,
    height=450
)

fig.show()


## **Datos de la tabla**

Esta tabla muestra la cantidad de videos en tendencias en YouTube, agrupados por categoría de contenido y por región. Las columnas representan las regiones (France, India, Japan, Russia y United States), y las filas muestran las diferentes categorías de contenido. Los números en las celdas indican la cantidad de videos en tendencia para cada combinación de categoría y región.

La categoría "Entertainment" tiene una cantidad significativamente mayor de videos en tendencias en todas las regiones, lo que sugiere que el contenido de entretenimiento es el más popular y ampliamente consumido.

- En India, la categoría "Entertainment" lidera con 32,924 videos, lo que refleja una gran preferencia por este tipo de contenido en esa región.
- En Japan, aunque "Entertainment" también es dominante, categorías como "Gaming" y "Film & Animation" tienen una representación notable, indicando una preferencia cultural por este tipo de contenido.
- Russia tiene un mayor equilibrio entre varias categorías como "Comedy" y "Film & Animation".
- United States muestra una representación alta tanto en "Entertainment" como en "Howto & Style".

Categorías como "Autos & Vehicles" tienen cantidades mucho menores en todas las regiones, lo que indica un interés más limitado por este tipo de contenido en tendencias.

Las diferencias en la cantidad de videos por categoría y región reflejan preferencias culturales, intereses de las audiencias, y posiblemente la influencia de los algoritmos de la plataforma en cada país.
La dominancia de ciertas categorías como "Entertainment" sugiere un enfoque global hacia contenido ampliamente atractivo, mientras que las variaciones regionales destacan oportunidades para contenido más específico y localizado.

# **Conclusión General del Proyecto**
El análisis global de los datos de tendencias en YouTube nos proporciona un panorama claro sobre cómo las diferentes regiones y categorías de contenido influyen en las tendencias globales y locales. 

Las categorías de "Entertainment" y "Music" dominan de manera consistente en cantidad de videos y proporción porcentual en todas las regiones.
Este dominio sugiere una preferencia generalizada por contenido atractivo, comercial y de amplio consumo.
Categorías como "Comedy", "Film & Animation", y "Gaming" también tienen representación importante, aunque con menor impacto que las categorías principales.

- Estados Unidos (23.8%) lidera en términos de videos tendencia, posiblemente debido a su gran volumen de producción y consumo.
- India y Francia tienen una fuerte representación de categorías como "Entertainment" y "Music", reflejando la riqueza cultural y el alto consumo digital en estas regiones.
- Japón, aunque con menor proporción (10.7%), destaca en categorías como "Film & Animation", indicando preferencias específicas y relevantes localmente.
- Rusia tiene una mayor representación en categorías como "News & Politics", lo que podría reflejar un interés más alto en temas de actualidad.

Categorías como "Autos & Vehicles", "Nonprofits & Activism", y "Science & Technology" tienen una representación baja, tanto en cantidad de videos como en proporción porcentual.
Esto podría deberse a una menor demanda de estos contenidos o a un sesgo en los algoritmos de tendencia hacia categorías de mayor atractivo comercial.

Los gráficos sugieren cierta estacionalidad en las tendencias, con picos y caídas regulares que podrían estar relacionados con eventos globales, lanzamientos importantes o campañas específicas, estas variaciones son particularmente notables en las categorías de "Entertainment" y "Music", que parecen reaccionar más a cambios externos.

Las proporciones relativamente estables de las categorías principales a lo largo del tiempo podrían indicar que los algoritmos de tendencia priorizan contenido de amplio interés y atractivo comercial.
Este enfoque deja menos visibilidad para categorías más específicas, como "Science & Technology" o "Education", lo que podría limitar la diversidad del contenido en tendencias.

## Recomendaciones Estratégicas
Las diferencias claras entre regiones sugieren que las estrategias de contenido deben adaptarse a los intereses y preferencias culturales específicos de cada región.

Aunque las categorías menores tienen menor representación, podrían ser nichos con oportunidades para el crecimiento, especialmente si se vinculan a intereses específicos de audiencias regionales.

Las plataformas y creadores de contenido pueden maximizar su impacto al sincronizar lanzamientos importantes con eventos globales o temporadas específicas.

# **Conclusión Final**
El análisis revela un claro dominio de contenido popular como "Entertainment" y "Music", con una representación significativa en todas las regiones. Sin embargo, también destaca la importancia de adaptar las estrategias de contenido a las preferencias regionales y explorar nichos con menor representación. Este enfoque puede ayudar a equilibrar la diversidad en las tendencias y ampliar la relevancia del contenido en audiencias globales y locales.

Las plataformas y los creadores de contenido tienen una oportunidad única de utilizar estos datos para optimizar sus estrategias y maximizar su alcance en mercados diversos y culturalmente ricos.